noir-projects/noir-contracts/contracts/nft_contract/src/main.nr

```diff
@@ -0,0 +1,191 @@
+// Minimal NFT implementation with `AuthWit` support that allows minting in public-only and transfers in both public
+// and private.
+contract NFT {
+    #[aztec(event)]
+    struct Transfer {
+        from: AztecAddress,
+        to: AztecAddress,
+        token_id: Field,
+    }
+
+    #[aztec(storage)]
+    struct Storage {
+        // The symbol of the NFT
+        symbol: SharedImmutable<FieldCompressedString>,
+        // The name of the NFT
+        name: SharedImmutable<FieldCompressedString>,
+        // The admin of the contract
+        admin: PublicMutable<AztecAddress>,
+        // Addresses that can mint
+        minters: Map<AztecAddress, PublicMutable<bool>>,
+        // Contains the NFTs owned by each address in private.
+        private_nfts: Map<AztecAddress, PrivateSet<NFTNote>>
+        // Contains number of NFTs owned by each address in public.
+        public_balances: Map<AztecAddress, PublicMutable<U128>>,
+        // A map from token ID to a boolean indicating if the NFT exists.
+        nft_exists: Map<Field, PublicMutable<bool>>,
```

Comment:
> If instead of having this we emitted a nullifier for each ID, then we could mint in private directly without even having to queue public function calls.

This is a good idea. I can imagine that there will be situations where you want it to be publicly known what nfts were minted but you could always just brute-force all the NFT nullifiers and see (assuming you are incrementing token-id counter).


> it'd be great if we could take the state variable and inject some param into it (e.g. in this case an npk)

The easiest thing here would be to just store 1 npk and use that for all NFTs. But what we could also do is to have 1 npk for a given range of token ids. Then we could expose a function where an admin could specify what range can use what npk (and for a given range an npk could be set only once):

`
    /// range_id - an id of a range - each range is 1000 NFTs long and id 0 is for NFTs with token ids ranging from 0 to 999, for id 1 for 1000-1999 etc.
    /// npk_m_hash - hash of the master nullifier public key which can be used to mint NFTs in a given range
    #[aztec(public)]
    fn register_npk(range_id: Field, npk_m_hash: Field) {
        assert(storage.minters.at(context.msg_sender()).read(), "caller is not minter");
        
        // store npk for the range in storage (probably in something like a map of shared immutables)
        ...
        
        // emit nullifier for a range
        let range_nullifier = hash([range_id], range_separator);
        context.push_nullifier(range_nullifier);
    }
`

With this approach we could essentially allow multiple minters to use different npk and we could also essentially achieve the effect of key rotation --> if npk got leaked a minter could just use a different range. (to make the interaction with admins unnecessary when "rotating" nullifier keys we could go even further and give power to a minter to set npk for a certain kind of ranges).

The only disadvantage would be that the token_id range would not be continuous but that does not seem like a big deal to me.

WDYT?

Reply:
> For now I'd keep this simple and not even have the npk.
Agree. IERC721 does not describe how tokens are minted as well so I would say it makes sense later on to come up with different kinds of "IERC721" extensions that implement different kinds of minting flows etc.